### PR TITLE
Initial logic for signal input migration

### PR DESCRIPTION
**Pull request by devversion**
This commit adds the code for the signal input migration. We've prototyped the migration and made it batchable— so that it can run for larger projects (e.g. using Beam); or in practice all of Google3 (via some internal tool).

The migration is currently still a prototype and is in progress of being finalized. In short the migration:

- Migrates `@Input()` declarations
- Migrates references to inputs in templates
- Migrates references to inputs in host bindings
- Migrates TypeScript references to inputs in the whole project.
   - It also creates temporary variables where needed;possible to support TS narrowing.

### CODE REVIEW

File: packages/core/schematics/migrations/signal-migration/src/passes/5_migrate_ts_references.ts [hunk 1]
```diff
@@ -0,0 +1,153 @@
...
+ *     this.input.charAt(0);
+ *   }
+ * }
+ * ```
+ *
+ * will be transformed into:
+ *
+ * ```
+ * someMethod() {
+ *   const input_1 = this.input()l
```
Comment (otodockal) [inline packages/core/schematics/migrations/signal-migration/src/passes/5_migrate_ts_references.ts:37]:
```suggestion
 *   const input_1 = this.input();
```

File: packages/core/schematics/migrations/signal-migration/src/flow_analysis/index.ts [hunk 2]
```diff
@@ -0,0 +1,284 @@
...
+import {getControlFlowContainer, isControlFlowBoundary} from './flow_containers';
+import {getFlowNode, traverseFlowForInterestingNodes} from './flow_node_traversal';
+import {unwrapParent} from '../utils/unwrap_parent';
+import assert from 'assert';
+
+/**
+ * Type describing an index of a node inside a control flow
+ * container.
+ */
+export type ControlFlowNodeIndex = number;
```
Comment (crisbeto) [inline packages/core/schematics/migrations/signal-migration/src/flow_analysis/index.ts:20]: nit: do we want a brand for this?

File: packages/core/schematics/migrations/signal-migration/src/input_detection/known_inputs.ts [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+  /** Maps classes to their directive info. */
+  private _classToDirectiveInfo = new Map<ts.ClassDeclaration, DirectiveInfo>();
+
+  /** Whether the given input exists. */
+  has(descr: Pick<InputDescriptor, 'key'>): boolean {
+    return this.knownInputIds.has(descr.key);
+  }
+
+  /** Whether the given class contains `@Input`s. */
+  isInputContainingClass(clazz: ts.ClassDeclaration): boolean {
```
Comment (crisbeto) [inline packages/core/schematics/migrations/signal-migration/src/input_detection/known_inputs.ts:53]: nit: the name seems "inverted" to me. Maybe something like `isInputOfClass`?

File: packages/core/schematics/migrations/signal-migration/src/input_detection/template_reference_visitor.ts [hunk 4]
```diff
@@ -0,0 +1,422 @@
...
+  expressionVisitor: TemplateExpressionReferenceVisitor<TmplAstNode>;
+  result: TmplInputExpressionReference<TmplAstNode>[] = [];
+
+  /**
+   * Whether we are currently descending into HTML AST nodes
+   * where bound attributes should be considered "narrowed".
+   *
+   * TODO: Remove with: https://github.com/angular/angular/pull/55456.
+   */
+  isAttributeInsideNarrowingExpression = false;
```
Comment (crisbeto) [inline packages/core/schematics/migrations/signal-migration/src/input_detection/template_reference_visitor.ts:75]: Should this be private?

File: packages/core/schematics/migrations/signal-migration/src/input_detection/template_reference_visitor.ts [hunk 5]
```diff
@@ -0,0 +1,422 @@
...
+  override visitLiteralMap(ast: LiteralMap, context: any) {
+    for (const [idx, key] of ast.keys.entries()) {
+      this.isInsideObjectShorthandExpression = !!key.isShorthandInitialized;
+      (ast.values[idx] as AST).visit(this, context);
+      this.isInsideObjectShorthandExpression = false;
+    }
+    super.visitLiteralMap(ast, context);
+  }
+
+  override visitPropertyRead(ast: PropertyRead) {
```
Comment (crisbeto) [inline packages/core/schematics/migrations/signal-migration/src/input_detection/template_reference_visitor.ts:255]: Do we need to account for safe reads/writes in this visitor as well?

File: packages/core/schematics/migrations/signal-migration/src/migration_host.ts [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+    return this._sourceFiles.has(file);
+  }
+
+  /** Retrieves a unique serializable ID for the given source file or file path. */
+  fileToId(file: ts.SourceFile | string): string {
+    if (typeof file !== 'string') {
+      // Assume that declaration files may appear in different workers,
+      // and in practice e.g. the input is actually part of a `.ts` file.
+      if (file.isDeclarationFile) {
+        file = file.fileName.replace(/\.d\.ts/, '.ts');
```
Comment (crisbeto) [inline packages/core/schematics/migrations/signal-migration/src/migration_host.ts:40]: nit: should we add `$` to the regex so it's a bit more explicit about where we want to replace?

File: packages/core/schematics/migrations/signal-migration/src/passes/2_find_source_file_references.ts [hunk 7]
```diff
@@ -0,0 +1,89 @@
...
+import {MigrationResult} from '../result';
+import {identifyHostBindingReferences} from './references/identify_host_references';
+import {identifyTemplateReferences} from './references/identify_template_references';
+import {identifyPotentialTypeScriptReference} from './references/identify_ts_references';
+import {PartialDirectiveTypeInCatalystTests} from '../pattern_advisors/partial_directive_type';
+import {InputReferenceKind} from '../utils/input_reference';
+
+/**
+ * Phase where we iterate through all source file references and
+ * detect references to inputs. These need to be migrated then.
```
Comment (crisbeto) [inline packages/core/schematics/migrations/signal-migration/src/passes/2_find_source_file_references.ts:25]: nit: this seems a bit unclear to me "These need to be migrated then."

**Review by crisbeto — APPROVED** [review]
Left some nits and questions, but otherwise LGTM

File: packages/core/schematics/migrations/signal-migration/src/create_program.ts [hunk 8]
```diff
@@ -0,0 +1,152 @@
...
+  templateTypeChecker.generateAllTypeCheckBlocks();
+
+  const {refEmitter, metaReader} = compiler['ensureAnalyzed']();
+  const typeChecker = userProgram.getTypeChecker();
+
+  const reflector = new TypeScriptReflectionHost(typeChecker);
+  const evaluator = new PartialEvaluator(reflector, typeChecker, null);
+  const dtsMetadataReader = new DtsMetadataReader(typeChecker, reflector);
+
+  const limitToRootNamesOnly = process.env['LIMIT_TO_ROOT_NAMES_ONLY'] === '1';
```
[the remark this answers is not in the document]
Comment (devversion) [reply packages/core/schematics/migrations/signal-migration/src/create_program.ts:120]: Probably fine like this. We are removing this at some point. It purely exists for testing right now.

File: packages/core/schematics/migrations/signal-migration/src/input_detection/known_inputs.ts (hunk 3 above)
Comment (devversion) [reply packages/core/schematics/migrations/signal-migration/src/input_detection/known_inputs.ts:53]: The helper is checking whether a class contains inputs. IIRC we need it that way for efficient lookups w/ inheritance

File: packages/compiler-cli/src/ngtsc/core/api/src/options.ts [hunk 9]
```diff
@@ -37,6 +37,12 @@ export interface TestOnlyOptions {
    */
   _enableTemplateTypeChecker?: boolean;
 
+  /**
+   * Whether components that are poisoned should still be processed.
+   * E.g. for generation of type check blocks and diagnostics.
+   */
+  _usePoisonedData?: boolean;
```
[the remark this answers is not in the document]
Comment (devversion) [reply packages/compiler-cli/src/ngtsc/core/api/src/options.ts:44]: Done

File: packages/core/schematics/migrations/signal-migration/src/convert-input/convert_to_signal.ts [hunk 10]
```diff
@@ -0,0 +1,158 @@
...
+  );
+
+  // TODO:
+  //   - modifiers (but private does not work)
+  //   - preserve custom decorators etc.
+
+  const result = ts.factory.createPropertyDeclaration(
+    undefined,
+    node.name,
+    undefined,
```
[the remark this answers is not in the document]
Comment (devversion) [reply packages/core/schematics/migrations/signal-migration/src/convert-input/convert_to_signal.ts:114]: The explicit types will be moved into `input<T>`, so no, explicit property types are intentionally avoided.

File: packages/core/schematics/migrations/signal-migration/src/flow_analysis/index.ts (hunk 2 above)
Comment (devversion) [reply packages/core/schematics/migrations/signal-migration/src/flow_analysis/index.ts:20]: I took a stab at this, but looks like this it's not worth the complexity for now; with regards to index types for example. It's mostly for documentation purposes.

File: packages/core/schematics/migrations/signal-migration/src/input_detection/template_reference_visitor.ts (hunk 5 above)
Comment (devversion) [reply packages/core/schematics/migrations/signal-migration/src/input_detection/template_reference_visitor.ts:255]: Adding a TODO. Good catch

File: packages/core/schematics/migrations/signal-migration/src/passes/2_find_source_file_references.ts (hunk 7 above)
Comment (devversion) [reply packages/core/schematics/migrations/signal-migration/src/passes/2_find_source_file_references.ts:25]: Clarified!

**Comment by devversion** [comment]
Thank you @crisbeto for the fast review 🎉

File: packages/core/schematics/migrations/signal-migration/src/convert-input/convert_to_signal.ts (hunk 10 above)
[the remark this answers is not in the document]
Comment (crisbeto) [reply packages/core/schematics/migrations/signal-migration/src/convert-input/convert_to_signal.ts:114]: Makes sense.

File: packages/core/schematics/migrations/signal-migration/src/input_detection/known_inputs.ts (hunk 3 above)
Comment (crisbeto) [reply packages/core/schematics/migrations/signal-migration/src/input_detection/known_inputs.ts:53]: I think the method is fine, it was the naming that was a bit confusing. Might just be me though.

**Review by crisbeto — APPROVED** [review]

**Comment by atscott** [comment]
This PR was merged into the repository by commit ad230630d622558ab33b73e1294f6f3768fe7907.

The changes were merged into the following branches: main

**Pull request closed without merging**